File: gist/gist.go
package gist

import (
	"bytes"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"time"

	"github.com/dramxx/later/config"
)

const baseURL = "https://api.github.com"

type GistResponse struct {
	Files map[string]struct {
		Content string `json:"content"`
	} `json:"files"`
}

type PatchRequest struct {
	Files map[string]struct {
		Content string `json:"content"`
	} `json:"files"`
}

func GetInbox(cfg *config.Config) (string, error) {
	url := fmt.Sprintf("%s/gists/%s", baseURL, cfg.Gist.GistID)
	req, err := http.NewRequest("GET", url, nil)
	if err != nil {
		return "", fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+cfg.Gist.Token)
	req.Header.Set("Accept", "application/vnd.github+json")

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return "", fmt.Errorf("network error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return "", fmt.Errorf("HTTP %d: failed to read body", resp.StatusCode)
		}
		return "", fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}

	var gistResp GistResponse
	if err := json.NewDecoder(resp.Body).Decode(&gistResp); err != nil {
		return "", fmt.Errorf("failed to parse response: %w", err)
	}

	file, ok := gistResp.Files["inbox.txt"]
	if !ok {
		return "", nil
	}

	return file.Content, nil
}

func UpdateInbox(cfg *config.Config, content string) error {
	url := fmt.Sprintf("%s/gists/%s", baseURL, cfg.Gist.GistID)

	patchReq := PatchRequest{
		Files: map[string]struct {
			Content string `json:"content"`
		}{
			"inbox.txt": {Content: content},
		},
	}

	jsonData, err := json.Marshal(patchReq)
	if err != nil {
		return fmt.Errorf("failed to marshal request: %w", err)
	}

	req, err := http.NewRequest("PATCH", url, bytes.NewBuffer(jsonData))
	if err != nil {
		return fmt.Errorf("failed to create request: %w", err)
	}

	req.Header.Set("Authorization", "Bearer "+cfg.Gist.Token)
	req.Header.Set("Accept", "application/vnd.github+json")
	req.Header.Set("Content-Type", "application/json")

	client := &http.Client{Timeout: 10 * time.Second}
	resp, err := client.Do(req)
	if err != nil {
		return fmt.Errorf("network error: %w", err)
	}
	defer resp.Body.Close()

	if resp.StatusCode != 200 {
		body, err := io.ReadAll(resp.Body)
		if err != nil {
			return fmt.Errorf("HTTP %d: failed to read body", resp.StatusCode)
		}
		return fmt.Errorf("HTTP %d: %s", resp.StatusCode, string(body))
	}

	return nil
}

File: src/gist.rs
use crate::config::Config;
use serde::{Deserialize, Serialize};

const BASE_URL: &str = "https://api.github.com";

#[derive(Deserialize)]
struct GistResponse {
    files: std::collections::HashMap<String, GistFile>,
}

#[derive(Deserialize)]
struct GistFile {
    content: String,
}

#[derive(Serialize)]
struct PatchRequest {
    files: std::collections::HashMap<String, PatchFile>,
}

#[derive(Serialize)]
struct PatchFile {
    content: String,
}

pub fn get_inbox(cfg: &Config) -> Result<String, Box<dyn std::error::Error>> {
    let url = format!("{}/gists/{}", BASE_URL, cfg.gist.gist_id);
    let client = reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()?;

    let response = client
        .get(&url)
        .header("User-Agent", "later-cli")
        .header("Authorization", format!("Bearer {}", cfg.gist.token))
        .header("Accept", "application/vnd.github+json")
        .send()?;

    let status = response.status();
    if !status.is_success() {
        let body = response.text()?;
        return Err(format!("HTTP {}: {}", status, body).into());
    }

    let gist: GistResponse = response.json()?;
    Ok(gist
        .files
        .get("inbox.txt")
        .map(|f| f.content.clone())
        .unwrap_or_default())
}

pub fn update_inbox(cfg: &Config, content: &str) -> Result<(), Box<dyn std::error::Error>> {
    let url = format!("{}/gists/{}", BASE_URL, cfg.gist.gist_id);

    let mut files = std::collections::HashMap::new();
    files.insert(
        "inbox.txt".to_string(),
        PatchFile {
            content: content.to_string(),
        },
    );

    let patch = PatchRequest { files };

    let client = reqwest::blocking::Client::builder()
        .timeout(std::time::Duration::from_secs(10))
        .build()?;

    let response = client
        .patch(&url)
        .header("User-Agent", "later-cli")
        .header("Authorization", format!("Bearer {}", cfg.gist.token))
        .header("Accept", "application/vnd.github+json")
        .header("Content-Type", "application/json")
        .json(&patch)
        .send()?;

    let status = response.status();
    if !status.is_success() {
        let body = response.text()?;
        return Err(format!("HTTP {}: {}", status, body).into());
    }

    Ok(())
}
